### ISSUE TITLE
globalObject: "this" work incorrect with ModuleFederation after 5.2.0

### ISSUE THREAD

**Issue by lomocc** [webpack/webpack#12031]
# Bug report

**What is the current behavior?**
the ModuleFederation crashed with `output.globalObject: "this" `

**If the current behavior is a bug, please provide the steps to reproduce.**
- clone `https://github.com/module-federation/module-federation-examples/tree/master/basic-host-remote`
- add `ModuleFederationPlugin`
- set `output.globalObject: "this" `
- set `package.json` (disable `arrowFunctions`)
```
"browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
```
- build and run

the output `remoteEntry.js` error:

```js
// this is undefined
var chunkLoadingGlobal = this["webpackChunkreact_app_shared"] = this["webpackChunkreact_app_shared"] || [];
```

**What is the expected behavior?**

`ModuleFederation` should works correct with: `output.globalObject: "this" `

**Other relevant information:**
webpack version:  5.2.0~5.5.1

**Comment by lomocc** [comment]
it works correct when set `output.globalObject: "self"`

**Comment by evilebottnawi** [comment]
Can't reproduce, all works fine, please create minimum reproducible test repo, both values works fine

**Comment by evilebottnawi** [comment]
I think it is the limitation, when you set `globalObject` webpack attach all runtime to this object, but in some cases it is not safe to do it, so it is why we use `self` by default, if you use `globalObject: 'something'` you will get problem too, what is use case to use `this`?

**Comment by lomocc** [comment]
1. [create-react-app](https://github.com/facebook/create-react-app/issues/7741) use `globalObject: this` for worker.
2. it works well at 5.1.3（with `arrowFunctions`）

>so I think this is a bug.

**Comment by evilebottnawi** [comment]
@lomocc No, it is not bug, it is limitation, `globalObject: this` will not work in workers, there is not `this` in web workers, you should use `self`

**Comment by evilebottnawi** [comment]
> it works well at 5.1.3（with arrowFunctions）

Yes, because arrow functions doesn't create context

**Comment by sokra** [comment]
It could be indeed a bug. Sounds like the bundle becomes strict mode only and `this` is undefined in strict mode.

As workaround avoid this as global object. Using the default should be fine instead.

**Comment by alexander-akait** [comment]
bump

**Comment by RReverser** [comment]
I'm running into an issue where `new Worker(new URL(...))` syntax used with a worker target that uses e.g.

```js
this.onmessage = console.log;
```

silently produces a broken worker that contains only

```js
console.log;
```

Is this the same issue or do I need to raise a separate one?

**Comment by alexander-akait** [comment]
@RReverser Can you provide more information? Need check

**Comment by RReverser** [comment]
Perhaps it's assuming that any Worker is ESM and hence strict by default, when it should be looking for `type: 'module'` to determine whether it's a script Worker, or a module Worker.

**Comment by sokra** [comment]
`this` in modules (e. g. CommonJs or ESM) is not the same as `this` in a script.

`this.onmessage = console.log;` this is equal to `exports.onmessage = console.log;` in CommonJs or `undefined.onmessage = console.log;` in ESM. So that won't work.

**Comment by RReverser** [comment]
@sokra I'm aware it's not the same. That's what I'm saying - that Webpack seems to assume that Worker is always a module, whereas by default in browsers it's a regular script and `this` is just a global object.

**Comment by Edge00** [comment]
My case:
First i found complied code contains **arrow fucntion**. so i add `target: ['web', 'es5']` to solve the problem.
Then code became:
```js
! function () {
  "use strict";
  /* `this` is undefined in strict mode */
  n = this.webpackChunkFoo = this.webpackChunkFoo || []
}
```
So i need to add `output.globalObject: 'window'` to solve this problem.

But it seems a little weird because once you set `target: ['web', 'es5']`. You need to configure `output.globalObject: 'window'` too. Which is not mentioned in documentation.

**Comment by vankop** [comment]
@alexander-akait what should we do here? sounds like this is limitation to use `globalObject: "this"`

**Comment by alexander-akait** [comment]
@vankop I think we should not use arrow function in some cases, but yes, it can be tricky to analyze (maybe regexp or warning?)

**Comment by vankop** [comment]
> I think we should not use arrow function in some cases

we use it only if environment supports it.

**Comment by RReverser** [comment]
Still an issue probably.
